### ISSUE TITLE
whitelist and private-dev combination

### ISSUE THREAD

**Issue by weeshy** [netblue30/firejail#608]
I really like idea of clean /dev directory, but I need access to some devices, that is not added by deafult with --private-dev option. I tried to whitelist them, and get empty /dev directory as result.

It would be great, if that options would cooperate with less devastation.

**Comment by Fred-Barclay** [comment]
Can you give an example so we can see what you did?
I suspect that you might need to `noblacklist` them too but that's just speculation at this point. ;)

**Comment by weeshy** [comment]
example: I want private /dev, but I need /dev/nvidia0 /dev/nvidiactl /dev/nvidia-modeset for optirun to work correctly.

**Comment by Fred-Barclay** [comment]
Sorry--what I mean was like `firejail --private-dev whitelist=/some/odd/directory` or whatever command you used. :wink:

**Comment by weeshy** [comment]
firejail --private-dev --whitelist=/dev/nvidia0

It gives empty /dev/ directory

**Comment by Fred-Barclay** [comment]
Try this instead:
`firejail --whitelist=/dev/nvidia0 --whitelist=/dev/nvidiactl --whitelist=/dev/nvidia-modeset`

These three directories will be the only ones visible in /dev/.

After some testing it appears that whitelist and private-dev are mutually exclusive; i.e. whitelist overrides private-dev. I'm not sure why this is (besides the nature of a whitelist being what it is) but I rarely use the private-* options. Maybe someone else with more experience can explain the technical details and/or correct where I'm wrong?

**Comment by netblue30** [comment]
The solution is to bring in private-dev all the nvidia devices. So far I have /dev/nvidia[0-9], /dev/nvidiactl, /dev/nvidia-modset and /dev/nvidia-uvm. I'll also bring in /dev/snd for sound. If I am missing something please let me know.

**Comment by weeshy** [comment]
firejail --whitelist works as intented: I just manually specified all devices that are normally is visible under private-dev and add nvidia devices to the list

**Comment by netblue30** [comment]
Yes, --whitelist will fix it for now, but I still have to bring in by default nvidia drivers in --private-dev. Anyway, thanks for bringing up this issue.

**Comment by netblue30** [comment]
I put the nvidia/private-dev thing on my todo list for now.

**Comment by nextime** [comment]
In debian this issue brings some issues. As in debian /dev/shm is a symlink to /run/shm, if i use a list of --whitelist to reproduce the behavior  of --private-dev i get an error on /dev/shm, and as i need also /dev/shm other than /dev/nvidia*, this seems a stopping issue for me

**Comment by netblue30** [comment]
Looking into it, thanks.

**Comment by nextime** [comment]
Just as a work-around in case anyone fall in this issue before a proper fix is found/implemented, i actually use this in /etc/rc.local:

rm  /dev/shm
mkdir /dev/shm
mount --bind /run/shm /dev/shm

This way we override the debian/devuan default of having /dev/shm as a synlink by actively remove it and recreate it as a bind mount of /run/shm, and then firejail works with --whitelist /dev/shm as intended.

**Comment by nextime** [comment]
UPDATE: same issue in debian and devuan apply also to things in /var/run ( for example, to whitelist /var/run/pcscd to permit use of smartcard authentication ) as /var/run is a symlink to /run

Same workaround as for /dev/shm ( so, remove the symlink and bind mount /run in /var/run in rc.local ) works.

**Comment by chiraag-nataraj** [comment]
FWIW, we have `--keep-dev-shm` now, which keeps the original `/dev/shm`. I don't know if that solves the issue given here though. And yes, `--keep-dev-shm` is compatible with `--private-dev`.

**Comment by chiraag-nataraj** [comment]
Closed due to inactivity. @nextime feel free to re-open if the workaround listed here doesn't work.
